nodejs/aws-serverless/cognito/userPool.ts

```diff
@@ -0,0 +1,366 @@
+        //
+        // If this attribute is not returned, it is assumed to be "false".
+        forceAliasCreation: boolean;
+    };
+}
+
+interface VerifyAuthChallengeEvent extends CommonUserPoolEvent {
+    triggerSource: "VerifyAuthChallengeResponse_Authentication";
+
+    request: {
+        userAttributes: Record<string, string>;
+
+        // This parameter comes from the Create Auth Challenge trigger, and is compared against a
+        // user’s challengeAnswer to determine whether the user passed the challenge.
+        privateChallengeParameters: Record<string, string>;
+
+        // The answer from the user's response to the challenge
+        challengeAnswer: Record<string, string>;
+    };
+
+    response: {
+        // Set to true if the user has successfully completed the challenge, or false otherwise.
+        answerCorrect: boolean;
+    };
+}
+
+interface CustomMessageEvent extends CommonUserPoolEvent {
+    triggerSource: "CustomMessage_SignUp" | "CustomMessage_AdminCreateUser" | "CustomMessage_ResendCode" | "CustomMessage_ForgotPassword" | "CustomMessage_UpdateUserAttribute" | "CustomMessage_VerifyUserAttribute" | "CustomMessage_Authentication";
+
+    request: {
+        userAttributes: Record<string, string>;
+
+        // A string for you to use as the placeholder for the verification code in the custom message.
+        codeParameter: string;
+
+        // The username parameter. It is a required request parameter for the admin create user flow.
+        usernameParameter: string;
+    };
+
+    response: {
+        // The custom SMS message to be sent to your users. Must include the codeParameter value
+        // received in the request.
+        smsMessage: string;
+
+        // The custom email message to be sent to your users. Must include the codeParameter value
+        // received in the request.
+        emailMessage: string;
+
+        // The subject line for the custom message.
+        emailSubject: string;
+    };
+}
+
+type UserPoolLambdaConfig = cognito.UserPoolArgs["lambdaConfig"];
+
+interface UserPoolHandlers {
+    createAuthChallenge?: Handler<CreateAuthChallengeEvent, CreateAuthChallengeEvent>;
+    customMessage?: Handler<CustomMessageEvent, CustomMessageEvent>;
+    defineAuthChallenge?: Handler<DefineAuthChallengeEvent, DefineAuthChallengeEvent>;
+    postAuthentication?: Handler<PostAuthenticationEvent, PostAuthenticationEvent>;
+    postConfirmation?: Handler<PostConfirmationEvent, PostConfirmationEvent>;
+    preAuthentication?: Handler<PreAuthenticationEvent, PreAuthenticationEvent>;
+    preSignUp?: Handler<PreSignUpEvent, PreSignUpEvent>;
+    preTokenGeneration?: Handler<PreTokenGenerationEvent, PreTokenGenerationEvent>;
+    userMigration?: Handler<UserMigrationEvent, UserMigrationEvent>;
+    verifyAuthChallengeResponse?: Handler<VerifyAuthChallengeEvent, VerifyAuthChallengeEvent>;
+}
+
+function createLambdaConfig(handlers: UserPoolHandlers | undefined): UserPoolLambdaConfig | undefined {
```

Comment:
@lukehoban Here's the other place we use this pattern.  We have a strongly typed, prescriptive JS object that you can fill out.  You then call createLambdaConfig on it to get the right object shape to pass into a userPool as the .lambdaConfig property value when creating it.

Reply:
in general, the pattern has been: createXXX where XXX is hte name of the property you need to provide to the resource's constructor.  so "createLambdaConfig" because there is a "lambdaConfig" property on UserPoolArgs.  Similarly, createLambdaFunctionAssociations because there is a property called lambdaFunctionAssociations in the cache properties in DistributionArgs.